### PR TITLE
feat(workflow): Unified Metric Rule API - POST & PUT

### PR DESCRIPTION
**Pull request by iProgramStuff**
This PR is intended to allow creation or modification of a metric alert rule - including it's triggers, and their actions - all in one call.

It does this by introducing a new serializer, UnifiedAlertRuleSerializer that adds the `triggers` fields. It also adds the actions field to the AlertRuleTriggerSerializer.   The `id` field has also been added to triggers and actions so that the FE can send them back for the put requests. 

UnifiedAlertRuleSerializer implements update, create, and validate functions.

We also now serialize the actions externally in the AlertRuleTriggerSerializer, so that the FE has this data available to show on the listing index page and it is also returned in the unified create/edit calls.

It adds backend constraints on the relationships between triggers and their threshold types and values in the validate function. I believe Billy will be implementing these constraints on the FE.

Tests included that send various valid and invalid payloads to the API. They also include a few tests that update a trigger's type and a trigger's threshold value.

### CODE REVIEW

File: tests/sentry/incidents/endpoints/test_organization_alert_rule_index.py [hunk 1]
```diff
@@ -89,6 +90,48 @@ def test_simple(self):
         alert_rule = AlertRule.objects.get(id=resp.data["id"])
         assert resp.data == serialize(alert_rule, self.user)
 
+    def test_unified_create(self):
```
Comment (dashed) [inline tests/sentry/incidents/endpoints/test_organization_alert_rule_index.py]: Good test.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 2]
```diff
@@ -33,15 +45,72 @@ def post(self, request, organization):
...
 
-        if serializer.is_valid():
-            alert_rule = serializer.save()
-            return Response(serialize(alert_rule, request.user), status=status.HTTP_201_CREATED)
+                if action_serializer.is_valid():
+                    try:
+                        action = action_serializer.save()
+                    except InvalidTriggerActionError as e:
+                        # Could not create a trigger action
+                        # TODO: ROLL BACK ALL CREATIONS (Trigger(s), action(s) if any, and the alert rule)
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: You can just wrap all these creations in a single transaction and it'll handle the rollback for you on exception. We probably don't want to have transaction logic in an endpoint, so we could make a function in logic.py to handle this. Or possibly in the serializer.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 3]
```diff
@@ -33,15 +45,72 @@ def post(self, request, organization):
         """
         Create an alert rule
         """
+        # TODO: Can save all created objects in a list for a rollback in case there are any errors. Just pass list to rollback function.
+        # TODO: Use create_alert_rule, create_alert_rule_trigger, and create_alert_rule_trigger_action?
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: Using the serializers is probably fine here, since they handle all the validation for you.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 4]
```diff
@@ -33,15 +45,72 @@ def post(self, request, organization):
...
         if not features.has("organizations:incidents", organization, actor=request.user):
             raise ResourceDoesNotExist
 
-        serializer = AlertRuleSerializer(
-            context={"organization": organization, "access": request.access}, data=request.data
+        data = deepcopy(request.data)
+        # TODO: Possibly before creating the rule, or sometime - ensure at least one trigger is defined. Also ensure that each trigger has at least one action.
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: I don't know if we require all triggers to have an action

Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: It might make sense to have an overall `Serializer` that handles the creation of the entire rule, and enforces the minimum one trigger, and the threshold constraints, etc. It could make sense to just build this all into the `AlertRuleSerializer`, potentially? We could still use the other serializers to verify that specific data.

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 5]
```diff
@@ -33,15 +45,72 @@ def post(self, request, organization):
...
+                    try:
+                        action = action_serializer.save()
+                    except InvalidTriggerActionError as e:
+                        # Could not create a trigger action
+                        # TODO: ROLL BACK ALL CREATIONS (Trigger(s), action(s) if any, and the alert rule)
+                        return Response(e.message, status=status.HTTP_400_BAD_REQUEST)
 
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+        # If we get here, the alert rule, triggers, and actions all created successfully. So return that.
+        return Response(serialize(alert_rule, request.user), status=status.HTTP_201_CREATED)
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: Does this now need to return all the details we created? Probably worth checking with @billyvg

Comment (billyvg) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py]: What is our general API pattern here?

Comment (wedamija) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py]: We don't really have a pattern for this situation. Usually we create things individually and return those individual items. Just wasn't sure if you'd want to use the response to show the full rule once it's created?

Comment (billyvg) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py]: Unlikely, we generally redirect you back to the listing page after you create/edit something

Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py]: I think it's general web practice to return the object you created, right?

But we /could/ save some data and return just a 201 if this is the only use case there will ever be?

It does even seem to be our own standard to return objects after creation, and we have been returning all of this data + overhead with the multi-endpoint method anyways. I have no idea how much money/bandwidth we'd actually be saving if we made the 201, but if it's negligible, I say we return the object.

Comment (wedamija) [reply src/sentry/incidents/endpoints/organization_alert_rule_index.py]: Not really looking to save money/bandwidth, mostly just trying to reduce complexity, since if we want to serialize the whole tree then we need a detailed serializer to do that. I think it's probably fine to just serialize the `alert_rule` like we're doing here.

File: src/sentry/incidents/logic.py [hunk 6]
```diff
@@ -591,6 +600,146 @@ class AlertRuleNameAlreadyUsedError(Exception):
...
 
 
+def validate_unified_alert_rule(data):
+    """Performs validation on an alert rule's POSTed data
+    This includes ensuring there is either 1 or 2 triggers, which each have actions, and have proper thresholds set.
+    The critical trigger should both alert and resolve 'after' the warning trigger. """
+    triggers = data.get("triggers", [])
+    if triggers:
+        if len(triggers) == 1:
+            if triggers[0]["label"] != CRITICAL_TRIGGER:
```
Comment (markstory) [inline src/sentry/incidents/logic.py]: Could you hit a KeyError here if label is not provided? Or will that constraint be enforced earlier in validation?

File: src/sentry/incidents/logic.py [hunk 7]
```diff
@@ -591,6 +600,146 @@ class AlertRuleNameAlreadyUsedError(Exception):
...
+        elif len(triggers) == 2:
+            if triggers[0]["label"] != CRITICAL_TRIGGER or triggers[1]["label"] != WARNING_TRIGGER:
+                return (
+                    False,
+                    "First trigger must be labeled 'CRITICAL', second trigger must be labeled 'WARNING'",
+                )
+            else:
+                # Verify thresholds. alertThreshold, resolveThreshold, threshholdType
+                if triggers[0]["thresholdType"] != triggers[1]["thresholdType"]:
+                    return False, "Must have matching threshold types (???)"
```
Comment (markstory) [inline src/sentry/incidents/logic.py]: Should `???` be replaced with something that will give the user more context on how to fix the validation error?

File: src/sentry/incidents/logic.py [hunk 8]
```diff
@@ -591,6 +600,146 @@ class AlertRuleNameAlreadyUsedError(Exception):
...
+
+    is_valid, message = validate_unified_alert_rule(data)
+    if not is_valid:
+        return {"error": True, "message": message}
+
+    from sentry.incidents.endpoints.serializers import (
+        AlertRuleSerializer,
+        AlertRuleTriggerSerializer,
+        AlertRuleTriggerActionSerializer,
+    )
```
Comment (markstory) [inline src/sentry/incidents/logic.py]: Why does this import need to be deferred?

Comment (iProgramStuff) [reply src/sentry/incidents/logic.py]: There seems to be some sort of circular import dependency error?

File: src/sentry/incidents/logic.py (hunk 7 above)
Comment (iProgramStuff) [reply src/sentry/incidents/logic.py]: Ah, I put it there because I wasn't sure if this was something to enforce. Now I am getting clarification on that.

File: src/sentry/incidents/logic.py (hunk 6 above)
Comment (iProgramStuff) [reply src/sentry/incidents/logic.py]: Label is sent by the frontend (not an input field by the user) as "CRITICAL"/"WARNING" atm. I think it is safe to assume proper API usage since it is a private API? Or do you think I should do key checks first?

File: src/sentry/incidents/endpoints/organization_alert_rule_index.py [hunk 9]
```diff
@@ -36,12 +38,15 @@ def post(self, request, organization):
         if not features.has("organizations:incidents", organization, actor=request.user):
             raise ResourceDoesNotExist
 
-        serializer = AlertRuleSerializer(
-            context={"organization": organization, "access": request.access}, data=request.data
-        )
+        data = deepcopy(request.data)
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_index.py]: Why do we deepcopy here?

File: src/sentry/incidents/logic.py [hunk 10]
```diff
@@ -591,6 +600,146 @@ class AlertRuleNameAlreadyUsedError(Exception):
...
+            "rule": The created `AlertRule`, # if 'error' is False, this is required.
+            "message": A unicode string representing the error encountered. Optional if 'error' is False, required if 'error' is True
+        }
+    """
+
+    is_valid, message = validate_unified_alert_rule(data)
+    if not is_valid:
+        return {"error": True, "message": message}
+
+    from sentry.incidents.endpoints.serializers import (
```
Comment (wedamija) [inline src/sentry/incidents/logic.py]: Don't import endpoint things into logic, that breaks our structure. Endpoints should only ever call business logic, not the other way around. We might need to restructure how the serializers work to get around this, or maybe just handle the transaction in a unified serializer that calls these sub serializers.

File: src/sentry/incidents/endpoints/serializers.py [hunk 11]
```diff
@@ -151,189 +348,110 @@ def update(self, instance, validated_data):
...
+        The critical trigger should both alert and resolve 'after' the warning trigger (whether that means > or < the value depends on threshold type/direction).
+        """
+        print("running validate.!?", data)
+        triggers = data.get("triggers", [])
+        if triggers:
+            if len(triggers) == 1:
+                if triggers[0].get("label", None) != CRITICAL_TRIGGER:
+                    raise serializers.ValidationError("First trigger must be labeled 'critical'")
+            elif len(triggers) == 2:
+                if triggers[0].get("label", None) != CRITICAL_TRIGGER or triggers[1]["label"] != WARNING_TRIGGER:
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: It might be easier to work with code from here onwards if you do
```
critical = triggers[0]
warning = triggers[1]
```

File: src/sentry/incidents/endpoints/serializers.py [hunk 12]
```diff
@@ -151,189 +348,110 @@ def update(self, instance, validated_data):
...
+                    raise serializers.ValidationError("First trigger must be labeled 'critical', second trigger must be labeled 'warning'")
+                else:
+                    if triggers[0]["thresholdType"] != triggers[1]["thresholdType"]:
+                        raise serializers.ValidationError("Must have matching threshold types (i.e. critical and warning triggers must both be an upper or lower bound)")
+
+                    if triggers[0]["thresholdType"] == 0:
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: I'd use `operator` here to reduce duplicate code.

```
import operator
if thresholdType == 0
    alert_op, resolve_op = operator.lt, operator.gt
else:
    alert_op, resolve_op = operator.gt, operator.lt

if alert_op(triggers[0]["alertThreshold], triggers[1]["alertThreshold"]):
    # Find a way to make the error message generic, or specify it when you assign the ops above
<etc>
```

Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: You should compare the thresholdType to the `Enum`

File: src/sentry/incidents/endpoints/serializers.py [hunk 13]
```diff
@@ -151,189 +348,110 @@ def update(self, instance, validated_data):
...
+            # Triggers have passed checks. Check that all triggers have at least one action now.
+            for trigger in triggers:
+                actions = trigger.get("actions", [])
+                if actions == []:
+                    raise serializers.ValidationError(trigger["label"] + " trigger must have an action.")
+        else:
+            raise serializers.ValidationError("Must include at least one trigger")
+
+        # return self._handle_old_fields_transition(data)
+        print("validated so far, gonna return ",super(UnifiedAlertRuleSerializer, self)._handle_old_fields_transition(data))
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: Probably don't need this anymore. If we're still passing multiple aggregations we can fix it on the FE as part of these changes.

File: src/sentry/incidents/endpoints/serializers.py [hunk 14]
```diff
@@ -151,189 +348,110 @@ def update(self, instance, validated_data):
...
+                    trigger_serializer = AlertRuleTriggerSerializer(
+                        context={
+                            "organization": self.context["organization"],
+                            "alert_rule": alert_rule,
+                            "access": self.context["access"],
+                        },
+                        data=trigger_data,
+                    )
+
+                    if trigger_serializer.is_valid():  # AlertRuleTriggerSerializer does the validation
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: If we get nested serializers working, it might make sense for the action data to be handled by the `AlertRuleTriggerSerializer`, since that data is more directly related there (if this is hard don't worry too much)

Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/serializers.py]: Nested serializers are working :)

I implemented get_attrs in AlertRuleTriggerSerializer that adds the actions to the serializer. now the frontend can also use them in the list display, and we can possibly add a "name"/"label" field to the actions like the issue alert ones have.

Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/serializers.py]: Hmm I wish the nested serializer could automatically call it's own update/create...and then I would also just drop the action_data loop into the AlertRuleTrigger's update/create. Might be a way.

File: src/sentry/api/serializers/models/alert_rule_trigger.py [hunk 15]
```diff
@@ -4,12 +4,27 @@
...
 
-from sentry.api.serializers import register, Serializer
-from sentry.incidents.models import AlertRuleTrigger, AlertRuleTriggerExclusion
+from sentry.api.serializers import register, serialize, Serializer
+from sentry.incidents.models import AlertRuleTrigger, AlertRuleTriggerAction, AlertRuleTriggerExclusion
 
 
 @register(AlertRuleTrigger)
 class AlertRuleTriggerSerializer(Serializer):
+    def get_attrs(self, item_list, user, **kwargs):
```
Comment (wedamija) [inline src/sentry/api/serializers/models/alert_rule_trigger.py]: I'd say these changes could be split into a separate pr. Let's just focus on the write path here.

File: src/sentry/api/serializers/rest_framework/base.py [hunk 16]
```diff
@@ -26,7 +26,9 @@ def convert_dict_key_case(obj, converter):
     Recursively converts the keys of a dictionary using the provided converter
     param.
     """
-    if not isinstance(obj, dict):
+    if isinstance(obj, list):
```
Comment (wedamija) [inline src/sentry/api/serializers/rest_framework/base.py]: Can you add a test to cover this case in `CamelSnakeModelSerializerTest`? (or just add a test for `convert_dict_key_case`)

File: src/sentry/incidents/endpoints/organization_alert_rule_details.py [hunk 17]
```diff
@@ -36,6 +35,7 @@ def put(self, request, organization, alert_rule):
 
     def delete(self, request, organization, alert_rule):
         try:
+            # TODO: Ensure delete_alert_rule deletes triggers + actions as well.
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/organization_alert_rule_details.py]: It should do since Django will follow references.

File: src/sentry/incidents/endpoints/serializers.py [hunk 18]
```diff
@@ -151,189 +353,117 @@ def update(self, instance, validated_data):
...
+                critical = triggers[0]
+                warning = triggers[1]
+                if critical.get('label', None) != CRITICAL_TRIGGER_LABEL or warning['label'] != WARNING_TRIGGER_LABEL:
+                    raise serializers.ValidationError('First trigger must be labeled "%s", second trigger must be labeled "%s"' % (CRITICAL_TRIGGER_LABEL, WARNING_TRIGGER_LABEL))
+                else:
+                    if critical['threshold_type'] != warning['threshold_type']:
+                        raise serializers.ValidationError('Must have matching threshold types (i.e. critical and warning triggers must both be an upper or lower bound)')
+
+                    if critical['threshold_type'] == AlertRuleThresholdType.ABOVE:
+                        alert_op, resolve_op = operator.lt, operator.lt
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: Ahh, didn't realize these were the same, we can just have op be a single variable then.

File: src/sentry/incidents/endpoints/serializers.py [hunk 19]
```diff
@@ -151,189 +353,117 @@ def update(self, instance, validated_data):
...
+                for actions_data in trigger_actions_data:
+                    create_alert_rule_trigger_action(trigger=trigger, **actions_data)
+            return alert_rule
+        except AlertRuleNameAlreadyUsedError:
+            raise serializers.ValidationError("This name is already in use for this project")
 
     def update(self, instance, validated_data):
         validated_data = self._remove_unchanged_fields(instance, validated_data)
-        return update_alert_rule_trigger_action(instance, **validated_data)
+        triggers_data = validated_data.pop("triggers")
```
Comment (wedamija) [inline src/sentry/incidents/endpoints/serializers.py]: Will we handle triggers/actions being removed here?

**Review by wedamija — COMMENTED** [review]
This is looking good. I'd say it probably makes sense to split out the frontend changes so we can just focus on the serializer changes here

File: src/sentry/api/serializers/rest_framework/base.py (hunk 16 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/rest_framework/base.py]: Added a test :)

File: src/sentry/incidents/endpoints/organization_alert_rule_details.py (hunk 17 above)
Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/organization_alert_rule_details.py]: That's what I thought & was hoping for. Just wanted to verify :-)

File: src/sentry/incidents/endpoints/serializers.py (hunk 18 above)
Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/serializers.py]: They weren't originally, but the logic was incorrect (eep). Fixed the logic.

File: src/sentry/incidents/endpoints/serializers.py (hunk 19 above)
Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/serializers.py]: Hmm...that's a good question. Or should we leave the individual delete for triggers/actions? 

Otherwise we'd need to do some magic on fetching the triggers/actions that weren't just updated/created or that exist and aren't in the list, etc..

File: src/sentry/api/serializers/models/alert_rule_trigger.py (hunk 15 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/models/alert_rule_trigger.py]: The tests actually rely on this. We use this serializer to get an object with triggers + actions (since using the model doesn't make this as easy).

File: src/sentry/incidents/endpoints/serializers.py (hunk 19 above)
Comment (iProgramStuff) [reply src/sentry/incidents/endpoints/serializers.py]: It's being handled now! Added a couple of tests for this as well.

**Review by wedamija — APPROVED** [review]
